Improved control over code changes (#925)
SINTEF/dlite: doc/contributors_guide/code_changes.md

Code changes
============

> Although that DLite haven't reached version 1.0.0, it strives to be stable and not introduce unnecessary backward-incompatible changes.

Reviewer: ```suggestion
Although that DLite has not reached version 1.0.0, the intention is for it to be stable and not introduce unnecessary backward-incompatible changes.
```

Versioning
----------
Before reaching version 1.0.0, DLite follows the following versioning rules that are well-known from Python:

* **patch release**: Only backward-compatible changes.
  - Addition of new backward-compatible features.

>   - New deprecation warnings can be added.

Reviewer: ```suggestion
  - New deprecation warnings can be added.
  - Bug fixes.
```

In C, use the `dlite_deprecation_warning()` function to mark a deprecation warning.

The idea with these functions is to make it easy to grep for deprecation warnings that should be remove in a new backward-incompatible release.
The following command can e.g. be used to find all deprecation warnings:

    find \( -name '*.py' -o -name '*.i' -o -name '*.c' \) ! -path './build*' | xargs grep -n -A1 deprecation_warning

How to handle different types of code changes
---------------------------------------------

### Backward-compatible API changes
The easiest API changes are those that only adds new functions or classes without changing the existing code, apart from adding deprecation warnings hinting about how to use the new API.

Adding new arguments to an existing function or new methods to existing classes falls also into this category.
**Note**, new (positional) arguments should be added to the end of the argument list, such that existing code using positional arguments will be unaffected by the change.

### New behavior

> All changes in behavior should be considered backward-incompatible.

Author: ```suggestion
All changes in behavior in existing code should be considered backward-incompatible.
```

Where it make sense, DLite will try to optionally keep the new/old behaviour over some releases in order for users to adapt to the new behavior.

Each optional new behavior should have a:
- name
- description
- release number for when the behavior was introduced (keeping old behavior as default)
- release number for when the new behavior should be the default (managed automatically)
- expected release number for when the old behavior should be removed (require developer effort)
- value: True means that the behavior is enabled.

> All behaviors are described in a single source file `src/dlite-behavior.c`.

Author: ```suggestion
All behavior changes will be described in a single source file `src/dlite-behavior.c`.
```
